extension of test class should NOT trigger no_magic_numbers rule

### New Issue Checklist

- [x ] Updated SwiftLint to the latest version
- [x ] I searched for [existing GitHub issues](https://github.com/realm/SwiftLint/issues)

### Describe the bug
- If extend from a test class, the no_magic_rule should not apply to the extension.
- https://github.com/realm/SwiftLint/issues/4896 that no_magic_number rules should not apply to Tests, that only works within the test class but not extension of the test class.

##### Complete output when running SwiftLint, including the stack trace and command used

```bash
$ swiftlint lint --path /Tests/LintTests/MyTests.swift  --no-cache --enable-all-rules
warning: The --path option is deprecated. Pass the path(s) to lint last to the swiftlint command.
Linting Swift files at paths Tests/LintTests/MyTests.swift
Linting 'MyTests.swift' (1/1)
warning: The `anyobject_protocol` rule is now deprecated and will be completely removed in a future release.
warning: The `inert_defer` rule is now deprecated and will be completely removed in a future release.
warning: The `unused_capture_list` rule is now deprecated and will be completely removed in a future release.
/Tests/LintTests/MyTests.swift:4:5: warning: Explicit ACL Violation: All declarations should specify Access Control Level keywords explicitly (explicit_acl)
/Tests/LintTests/MyTests.swift:5:13: warning: Explicit Type Interface Violation: Properties should have a type interface (explicit_type_interface)
/Tests/LintTests/MyTests.swift:12:13: warning: Explicit Type Interface Violation: Properties should have a type interface (explicit_type_interface)
/Tests/LintTests/MyTests.swift:11:5: warning: Missing Docs Violation: public declarations should be documented (missing_docs)
/Tests/LintTests/MyTests.swift:10:1: error: No Extension Access Modifier Violation: Prefer not to use extension access modifiers (no_extension_access_modifier)
/Tests/LintTests/MyTests.swift:10:8: warning: No Grouping Extension Violation: Extensions shouldn't be used to group code within the same source file (no_grouping_extension)

/Tests/LintTests/MyTests.swift:13:42: warning: No Magic Numbers Violation: Magic numbers should be replaced by named constants (no_magic_numbers)

/Tests/LintTests/MyTests.swift:6:9: warning: Prefer Nimble Violation: Prefer Nimble matchers over XCTAssert functions (prefer_nimble)
/Tests/LintTests/MyTests.swift:13:9: warning: Prefer Nimble Violation: Prefer Nimble matchers over XCTAssert functions (prefer_nimble)
/Tests/LintTests/MyTests.swift:3:14: warning: Required Deinit Violation: Classes should have an explicit deinit method (required_deinit)
Done linting! Found 10 violations, 1 serious in 1 file.
```

### Environment

* SwiftLint version: 0.52.4
* both homebrew and swiftcommandplugins: both use the same swiftlint version and can reproduce the bug
* Paste your configuration file:

```yml
opt_in_rules:
- no_magic_numbers
```
We use default settings so "XCTestCase" should by default ignored by no_magic_number rule

* Are you using [nested configurations](https://github.com/realm/SwiftLint#nested-configurations)?
no
* Which Xcode version are you using (check `xcodebuild -version`)?
Xcode 14.3 Build version 14E222b
* Do you have a sample that shows the issue? Run `echo "[string here]" | swiftlint lint --no-cache --use-stdin --enable-all-rules`
  to quickly test if your example is really demonstrating the issue. If your example is more
  complex, you can use `swiftlint lint --path [file here] --no-cache --enable-all-rules`.

```swift
import XCTest

final class MyTests: XCTestCase {
    func testMagicNumberRuleNotApplied() {
        let collection = []
        XCTAssertEqual(collection.count, 2) 
    }
}

extension MyTests {
    func testMagicNumberRuleApplied() {
        let collection = []
        XCTAssertEqual(collection.count, 2) // This triggers a no-magic-rule violation:
    }
}
```

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -71,6 +71,13 @@
   [Martin Redington](https://github.com/mildm8nnered)
   [#4876](https://github.com/realm/SwiftLint/issues/4876)
 
+* The `no_magic_numbers` rule will not trigger for violations in an
+  extension, if the extended class inherits from one of the specified
+  `test_parent_classes`, as long as the class declaration and the
+  extension are in the same source file.  
+  [Martin Redington](https://github.com/mildm8nnered)
+  [#5137](https://github.com/realm/SwiftLint/issues/5137)
+
 ## 0.52.4: Lid Switch
 
 #### Breaking

--- a/Source/SwiftLintBuiltInRules/Rules/Idiomatic/NoMagicNumbersRule.swift
+++ b/Source/SwiftLintBuiltInRules/Rules/Idiomatic/NoMagicNumbersRule.swift
@@ -9,80 +9,80 @@ struct NoMagicNumbersRule: SwiftSyntaxRule, OptInRule, ConfigurationProviderRule
         description: "Magic numbers should be replaced by named constants",
         kind: .idiomatic,
         nonTriggeringExamples: [
-            Example("var foo = 123"),
-            Example("static let bar: Double = 0.123"),
-            Example("let a = b + 1.0"),
-            Example("array[0] + array[1] "),
-            Example("let foo = 1_000.000_01"),
-            Example("// array[1337]"),
-            Example("baz(\"9999\")"),
-            Example("""
+            "var foo = 123",
+            "static let bar: Double = 0.123",
+            "let a = b + 1.0",
+            "array[0] + array[1] ",
+            "let foo = 1_000.000_01",
+            "// array[1337]",
+            "baz(\"9999\")",
+            """
             func foo() {
                 let x: Int = 2
                 let y = 3
                 let vector = [x, y, -1]
             }
-            """),
-            Example("""
+            """,
+            """
             class A {
                 var foo: Double = 132
                 static let bar: Double = 0.98
             }
-            """),
-            Example("""
+            """,
+            """
             @available(iOS 13, *)
             func version() {
                 if #available(iOS 13, OSX 10.10, *) {
                     return
                 }
             }
-            """),
-            Example("""
+            """,
+            """
             enum Example: Int {
                 case positive = 2
                 case negative = -2
             }
-            """),
-            Example("""
+            """,
+            """
             class FooTests: XCTestCase {
                 let array: [Int] = []
                 let bar = array[42]
             }
-            """),
-            Example("""
+            """,
+            """
             class FooTests: XCTestCase {
                 class Bar {
                     let array: [Int] = []
                     let bar = array[42]
                 }
             }
-            """),
-            Example("""
+            """,
+            """
             class MyTest: XCTestCase {}
             extension MyTest {
                 let a = Int(3)
             }
-            """),
-            Example("""
+            """,
+            """
             extension MyTest {
                 let a = Int(3)
             }
             class MyTest: XCTestCase {}
-            """)
+            """,
         ],
         triggeringExamples: [
-            Example("foo(↓321)"),
-            Example("bar(↓1_000.005_01)"),
-            Example("array[↓42]"),
-            Example("let box = array[↓12 + ↓14]"),
-            Example("let a = b + ↓2.0"),
-            Example("Color.primary.opacity(isAnimate ? ↓0.1 : ↓1.5)"),
-            Example("""
+            "foo(↓321)",
+            "bar(↓1_000.005_01)",
+            "array[↓42]",
+            "let box = array[↓12 + ↓14]",
+            "let a = b + ↓2.0",
+            "Color.primary.opacity(isAnimate ? ↓0.1 : ↓1.5)",
+            """
                     class MyTest: XCTestCase {}
                     extension NSObject {
                         let a = Int(↓3)
                     }
-            """)
+            """,
         ]
     )
 

--- a/Source/SwiftLintBuiltInRules/Rules/Idiomatic/NoMagicNumbersRule.swift
+++ b/Source/SwiftLintBuiltInRules/Rules/Idiomatic/NoMagicNumbersRule.swift
@@ -172,7 +172,8 @@ private extension NoMagicNumbersRule {
             guard let possibleViolationsForClass = possibleViolations[className] else {
                 return
             }
-            violations.removeAll { possibleViolationsForClass.contains($0) }
+            let violationsToRemove = Set(possibleViolationsForClass)
+            violations.removeAll { violationsToRemove.contains($0) }
             possibleViolations[className] = nil
         }
     }

--- a/Source/SwiftLintBuiltInRules/Rules/Idiomatic/NoMagicNumbersRule.swift
+++ b/Source/SwiftLintBuiltInRules/Rules/Idiomatic/NoMagicNumbersRule.swift
@@ -133,13 +133,13 @@ private extension NoMagicNumbersRule {
             guard !node.isMemberOfATestClass(testParentClasses) else {
                 return
             }
-            if let extendedClassName = node.extendedClassName() {
-                if testClasses.contains(extendedClassName) == false {
+            if let extendedTypeName = node.extendedTypeName() {
+                if testClasses.contains(extendedTypeName) == false {
                     violations.append(violation)
-                    if nonTestClasses.contains(extendedClassName) == false {
-                        var possibleViolationsForClass = possibleViolations[extendedClassName] ?? []
+                    if nonTestClasses.contains(extendedTypeName) == false {
+                        var possibleViolationsForClass = possibleViolations[extendedTypeName] ?? []
                         possibleViolationsForClass.insert(violation)
-                        possibleViolations[extendedClassName] = possibleViolationsForClass
+                        possibleViolations[extendedTypeName] = possibleViolationsForClass
                     }
                 }
             } else {
@@ -188,7 +188,7 @@ private extension ExprSyntaxProtocol {
         return false
     }
 
-    func extendedClassName() -> String? {
+    func extendedTypeName() -> String? {
         var parent = parent
         while parent != nil {
             if let extensionDecl = parent?.as(ExtensionDeclSyntax.self) {

--- a/Source/SwiftLintBuiltInRules/Rules/Idiomatic/NoMagicNumbersRule.swift
+++ b/Source/SwiftLintBuiltInRules/Rules/Idiomatic/NoMagicNumbersRule.swift
@@ -88,11 +88,11 @@ struct NoMagicNumbersRule: SwiftSyntaxRule, OptInRule, ConfigurationProviderRule
             Example("let a = b + ↓2.0"),
             Example("Color.primary.opacity(isAnimate ? ↓0.1 : ↓1.5)"),
             Example("""
-            extension AVAsset {
-                enum VideoOrientation {
+            extension Foo {
+                enum Orientation {
                     case right, up, left, down
 
-                    static func fromVideoWithAngle(ofDegree degree: CGFloat) -> VideoOrientation? {
+                    static func fromAngle(ofDegree degree: CGFloat) -> Orientation? {
                         switch Int(degree) {
                         case 0: return .right
                         case ↓90: return .up
@@ -128,7 +128,7 @@ private extension NoMagicNumbersRule {
             let className = node.identifier.text
             if node.isXCTestCase(testParentClasses) {
                 testClasses.insert(className)
-                processPossibleViolations(forClassName: className)
+                removeViolations(forClassName: className)
             } else {
                 nonTestClasses.insert(className)
             }
@@ -154,13 +154,13 @@ private extension NoMagicNumbersRule {
             guard !node.isMemberOfATestClass(testParentClasses) else {
                 return
             }
-            if let extendedTypeName = node.extendedClassname() {
-                if testClasses.contains(extendedTypeName) == false {
+            if let extendedClassName = node.extendedClassName() {
+                if testClasses.contains(extendedClassName) == false {
                     violations.append(violation)
-                    if nonTestClasses.contains(extendedTypeName) == false {
-                        var possibleViolationsForClass = possibleViolations[extendedTypeName] ?? []
+                    if nonTestClasses.contains(extendedClassName) == false {
+                        var possibleViolationsForClass = possibleViolations[extendedClassName] ?? []
                         possibleViolationsForClass.append(violation)
-                        possibleViolations[extendedTypeName] = possibleViolationsForClass
+                        possibleViolations[extendedClassName] = possibleViolationsForClass
                     }
                 }
             } else {
@@ -168,13 +168,11 @@ private extension NoMagicNumbersRule {
             }
         }
 
-        private func processPossibleViolations(forClassName className: String) {
+        private func removeViolations(forClassName className: String) {
             guard let possibleViolationsForClass = possibleViolations[className] else {
                 return
             }
-            if testClasses.contains(className) {
-                violations.removeAll { possibleViolationsForClass.contains($0) }
-            }
+            violations.removeAll { possibleViolationsForClass.contains($0) }
             possibleViolations[className] = nil
         }
     }
@@ -211,7 +209,7 @@ private extension ExprSyntaxProtocol {
         return false
     }
 
-    func extendedClassname() -> String? {
+    func extendedClassName() -> String? {
         var parent = parent
         while parent != nil {
             if let extensionDecl = parent?.as(ExtensionDeclSyntax.self) {

--- a/Source/SwiftLintBuiltInRules/Rules/Idiomatic/NoMagicNumbersRule.swift
+++ b/Source/SwiftLintBuiltInRules/Rules/Idiomatic/NoMagicNumbersRule.swift
@@ -68,7 +68,7 @@ struct NoMagicNumbersRule: SwiftSyntaxRule, OptInRule, ConfigurationProviderRule
                 let a = Int(3)
             }
             class MyTest: XCTestCase {}
-            """),
+            """)
         ],
         triggeringExamples: [
             Example("foo(↓321)"),
@@ -82,7 +82,7 @@ struct NoMagicNumbersRule: SwiftSyntaxRule, OptInRule, ConfigurationProviderRule
                     extension NSObject {
                         let a = Int(↓3)
                     }
-            """),
+            """)
         ]
     )
 

--- a/Source/SwiftLintBuiltInRules/Rules/Idiomatic/NoMagicNumbersRule.swift
+++ b/Source/SwiftLintBuiltInRules/Rules/Idiomatic/NoMagicNumbersRule.swift
@@ -9,80 +9,80 @@ struct NoMagicNumbersRule: SwiftSyntaxRule, OptInRule, ConfigurationProviderRule
         description: "Magic numbers should be replaced by named constants",
         kind: .idiomatic,
         nonTriggeringExamples: [
-            "var foo = 123",
-            "static let bar: Double = 0.123",
-            "let a = b + 1.0",
-            "array[0] + array[1] ",
-            "let foo = 1_000.000_01",
-            "// array[1337]",
-            "baz(\"9999\")",
-            """
+            Example("var foo = 123"),
+            Example("static let bar: Double = 0.123"),
+            Example("let a = b + 1.0"),
+            Example("array[0] + array[1] "),
+            Example("let foo = 1_000.000_01"),
+            Example("// array[1337]"),
+            Example("baz(\"9999\")"),
+            Example("""
             func foo() {
                 let x: Int = 2
                 let y = 3
                 let vector = [x, y, -1]
             }
-            """,
-            """
+            """),
+            Example("""
             class A {
                 var foo: Double = 132
                 static let bar: Double = 0.98
             }
-            """,
-            """
+            """),
+            Example("""
             @available(iOS 13, *)
             func version() {
                 if #available(iOS 13, OSX 10.10, *) {
                     return
                 }
             }
-            """,
-            """
+            """),
+            Example("""
             enum Example: Int {
                 case positive = 2
                 case negative = -2
             }
-            """,
-            """
+            """),
+            Example("""
             class FooTests: XCTestCase {
                 let array: [Int] = []
                 let bar = array[42]
             }
-            """,
-            """
+            """),
+            Example("""
             class FooTests: XCTestCase {
                 class Bar {
                     let array: [Int] = []
                     let bar = array[42]
                 }
             }
-            """,
-            """
+            """),
+            Example("""
             class MyTest: XCTestCase {}
             extension MyTest {
                 let a = Int(3)
             }
-            """,
-            """
+            """),
+            Example("""
             extension MyTest {
                 let a = Int(3)
             }
             class MyTest: XCTestCase {}
-            """,
+            """)
         ],
         triggeringExamples: [
-            "foo(↓321)",
-            "bar(↓1_000.005_01)",
-            "array[↓42]",
-            "let box = array[↓12 + ↓14]",
-            "let a = b + ↓2.0",
-            "Color.primary.opacity(isAnimate ? ↓0.1 : ↓1.5)",
-            """
+            Example("foo(↓321)"),
+            Example("bar(↓1_000.005_01)"),
+            Example("array[↓42]"),
+            Example("let box = array[↓12 + ↓14]"),
+            Example("let a = b + ↓2.0"),
+            Example("Color.primary.opacity(isAnimate ? ↓0.1 : ↓1.5)"),
+            Example("""
                     class MyTest: XCTestCase {}
                     extension NSObject {
                         let a = Int(↓3)
                     }
-            """,
+            """)
         ]
     )
 

--- a/Source/SwiftLintBuiltInRules/Rules/Idiomatic/NoMagicNumbersRule.swift
+++ b/Source/SwiftLintBuiltInRules/Rules/Idiomatic/NoMagicNumbersRule.swift
@@ -58,27 +58,17 @@ struct NoMagicNumbersRule: SwiftSyntaxRule, OptInRule, ConfigurationProviderRule
             }
             """),
             Example("""
-            class FooTests: XCTestCase {
-                let baz: [Int] = Array(repeating: 0, count: 10)
-            }
-
-            extension FooTests {
-                class Bar {
-                   let baz: [Int] = Array(repeating: 0, count: 10)
-                }
+            class MyTest: XCTestCase {}
+            extension MyTest {
+                let a = Int(3)
             }
             """),
             Example("""
-            extension FooTests {
-                class Bar {
-                   let baz: [Int] = Array(repeating: 0, count: 10)
-                }
+            extension MyTest {
+                let a = Int(3)
             }
-
-            class FooTests: XCTestCase {
-                let baz: [Int] = Array(repeating: 0, count: 10)
-            }
-            """)
+            class MyTest: XCTestCase {}
+            """),
         ],
         triggeringExamples: [
             Example("foo(↓321)"),
@@ -88,22 +78,11 @@ struct NoMagicNumbersRule: SwiftSyntaxRule, OptInRule, ConfigurationProviderRule
             Example("let a = b + ↓2.0"),
             Example("Color.primary.opacity(isAnimate ? ↓0.1 : ↓1.5)"),
             Example("""
-            extension Foo {
-                enum Orientation {
-                    case right, up, left, down
-
-                    static func fromAngle(ofDegree degree: CGFloat) -> Orientation? {
-                        switch Int(degree) {
-                        case 0: return .right
-                        case ↓90: return .up
-                        case ↓180: return .left
-                        case -↓90: return .down
-                        default: return nil
-                        }
+                    class MyTest: XCTestCase {}
+                    extension NSObject {
+                        let a = Int(↓3)
                     }
-                }
-            }
-            """)
+            """),
         ]
     )
 

--- a/Source/SwiftLintBuiltInRules/Rules/Idiomatic/NoMagicNumbersRule.swift
+++ b/Source/SwiftLintBuiltInRules/Rules/Idiomatic/NoMagicNumbersRule.swift
@@ -96,7 +96,7 @@ private extension NoMagicNumbersRule {
         private let testParentClasses: Set<String>
         private var testClasses: Set<String> = []
         private var nonTestClasses: Set<String> = []
-        private var possibleViolations: [String: Set<ReasonedRuleViolation>] = [:]
+        private var possibleViolations: [String: Set<AbsolutePosition>] = [:]
 
         init(viewMode: SyntaxTreeViewMode, testParentClasses: Set<String>) {
             self.testParentClasses = testParentClasses
@@ -145,9 +145,10 @@ private extension NoMagicNumbersRule {
         }
 
         private func removeViolations(forClassName className: String) {
-            guard let violationsToRemove = possibleViolations[className] else {
+            guard let possibleViolationsForClass = possibleViolations[className] else {
                 return
             }
+            let violationsToRemove = Set(possibleViolationsForClass.map { ReasonedRuleViolation(position: $0) })
             violations.removeAll { violationsToRemove.contains($0) }
             possibleViolations.removeValue(forKey: className)
         }

--- a/Source/SwiftLintBuiltInRules/Rules/Idiomatic/NoMagicNumbersRule.swift
+++ b/Source/SwiftLintBuiltInRules/Rules/Idiomatic/NoMagicNumbersRule.swift
@@ -136,9 +136,7 @@ private extension NoMagicNumbersRule {
                 if testClasses.contains(extendedTypeName) == false {
                     violations.append(violation)
                     if nonTestClasses.contains(extendedTypeName) == false {
-                        var possibleViolationsForClass = possibleViolations[extendedTypeName] ?? []
-                        possibleViolationsForClass.insert(violation)
-                        possibleViolations[extendedTypeName] = possibleViolationsForClass
+                        possibleViolations[extendedTypeName, default: []].insert(violation)
                     }
                 }
             } else {

--- a/Source/SwiftLintBuiltInRules/Rules/Idiomatic/NoMagicNumbersRule.swift
+++ b/Source/SwiftLintBuiltInRules/Rules/Idiomatic/NoMagicNumbersRule.swift
@@ -67,7 +67,7 @@ struct NoMagicNumbersRule: SwiftSyntaxRule, OptInRule, ConfigurationProviderRule
                    let baz: [Int] = Array(repeating: 0, count: 10)
                 }
             }
-            """),
+            """).focused(),
             Example("""
             extension FooTests {
                 class Bar {
@@ -86,7 +86,24 @@ struct NoMagicNumbersRule: SwiftSyntaxRule, OptInRule, ConfigurationProviderRule
             Example("array[↓42]"),
             Example("let box = array[↓12 + ↓14]"),
             Example("let a = b + ↓2.0"),
-            Example("Color.primary.opacity(isAnimate ? ↓0.1 : ↓1.5)")
+            Example("Color.primary.opacity(isAnimate ? ↓0.1 : ↓1.5)"),
+            Example("""
+            extension AVAsset {
+                enum VideoOrientation {
+                    case right, up, left, down
+
+                    static func fromVideoWithAngle(ofDegree degree: CGFloat) -> VideoOrientation? {
+                        switch Int(degree) {
+                        case 0: return .right
+                        case ↓90: return .up
+                        case ↓180: return .left
+                        case -↓90: return .down
+                        default: return nil
+                        }
+                    }
+                }
+            }
+            """)
         ]
     )
 
@@ -108,13 +125,13 @@ private extension NoMagicNumbersRule {
         }
 
         override func visitPost(_ node: ClassDeclSyntax) {
-            let className = node.trimmedDescription
+            let className = node.identifier.text
             if node.isXCTestCase(testParentClasses) {
                 testClasses.insert(className)
+                processPossibleViolations(forClassName: className)
             } else {
                 nonTestClasses.insert(className)
             }
-            processPossibleViolations(forClassName: className)
         }
 
         override func visitPost(_ node: FloatLiteralExprSyntax) {
@@ -138,12 +155,13 @@ private extension NoMagicNumbersRule {
                 return
             }
             if let extendedTypeName = node.extendedClassname() {
-                if testClasses.contains(extendedTypeName) {
+                if testClasses.contains(extendedTypeName) == false {
                     violations.append(violation)
-                } else if nonTestClasses.contains(extendedTypeName) == false {
-                    var possibleViolationsForClass = possibleViolations[extendedTypeName] ?? []
-                    possibleViolationsForClass.append(violation)
-                    possibleViolations[extendedTypeName] = possibleViolationsForClass
+                    if nonTestClasses.contains(extendedTypeName) == false {
+                        var possibleViolationsForClass = possibleViolations[extendedTypeName] ?? []
+                        possibleViolationsForClass.append(violation)
+                        possibleViolations[extendedTypeName] = possibleViolationsForClass
+                    }
                 }
             } else {
                 violations.append(violation)
@@ -154,8 +172,9 @@ private extension NoMagicNumbersRule {
             guard let possibleViolationsForClass = possibleViolations[className] else {
                 return
             }
-            violations.append(contentsOf: possibleViolationsForClass)
-            violations.sort(by: { $0.position < $1.position })
+            if testClasses.contains(className) {
+                violations.removeAll { possibleViolationsForClass.contains($0) }
+            }
             possibleViolations[className] = nil
         }
     }

--- a/Source/SwiftLintBuiltInRules/Rules/Idiomatic/NoMagicNumbersRule.swift
+++ b/Source/SwiftLintBuiltInRules/Rules/Idiomatic/NoMagicNumbersRule.swift
@@ -96,7 +96,7 @@ private extension NoMagicNumbersRule {
         private let testParentClasses: Set<String>
         private var testClasses: Set<String> = []
         private var nonTestClasses: Set<String> = []
-        private var possibleViolations: [String: [ReasonedRuleViolation]] = [:]
+        private var possibleViolations: [String: Set<ReasonedRuleViolation>] = [:]
 
         init(viewMode: SyntaxTreeViewMode, testParentClasses: Set<String>) {
             self.testParentClasses = testParentClasses
@@ -138,7 +138,7 @@ private extension NoMagicNumbersRule {
                     violations.append(violation)
                     if nonTestClasses.contains(extendedClassName) == false {
                         var possibleViolationsForClass = possibleViolations[extendedClassName] ?? []
-                        possibleViolationsForClass.append(violation)
+                        possibleViolationsForClass.insert(violation)
                         possibleViolations[extendedClassName] = possibleViolationsForClass
                     }
                 }
@@ -148,12 +148,11 @@ private extension NoMagicNumbersRule {
         }
 
         private func removeViolations(forClassName className: String) {
-            guard let possibleViolationsForClass = possibleViolations[className] else {
+            guard let violationsToRemove = possibleViolations[className] else {
                 return
             }
-            let violationsToRemove = Set(possibleViolationsForClass)
             violations.removeAll { violationsToRemove.contains($0) }
-            possibleViolations[className] = nil
+            possibleViolations.removeValue(forKey: className)
         }
     }
 }

--- a/Source/SwiftLintBuiltInRules/Rules/Idiomatic/NoMagicNumbersRule.swift
+++ b/Source/SwiftLintBuiltInRules/Rules/Idiomatic/NoMagicNumbersRule.swift
@@ -56,6 +56,28 @@ struct NoMagicNumbersRule: SwiftSyntaxRule, OptInRule, ConfigurationProviderRule
                     let bar = array[42]
                 }
             }
+            """),
+            Example("""
+            class FooTests: XCTestCase {
+                let baz: [Int] = Array(repeating: 0, count: 10)
+            }
+
+            extension FooTests {
+                class Bar {
+                   let baz: [Int] = Array(repeating: 0, count: 10)
+                }
+            }
+            """),
+            Example("""
+            extension FooTests {
+                class Bar {
+                   let baz: [Int] = Array(repeating: 0, count: 10)
+                }
+            }
+
+            class FooTests: XCTestCase {
+                let baz: [Int] = Array(repeating: 0, count: 10)
+            }
             """)
         ],
         triggeringExamples: [
@@ -76,22 +98,65 @@ struct NoMagicNumbersRule: SwiftSyntaxRule, OptInRule, ConfigurationProviderRule
 private extension NoMagicNumbersRule {
     final class Visitor: ViolationsSyntaxVisitor {
         private let testParentClasses: Set<String>
+        private var testClasses: Set<String> = []
+        private var nonTestClasses: Set<String> = []
+        private var possibleViolations: [String: [ReasonedRuleViolation]] = [:]
 
         init(viewMode: SyntaxTreeViewMode, testParentClasses: Set<String>) {
             self.testParentClasses = testParentClasses
             super.init(viewMode: viewMode)
         }
 
-        override func visitPost(_ node: FloatLiteralExprSyntax) {
-            if node.isMemberOfATestClass(testParentClasses) == false, node.floatingDigits.isMagicNumber {
-                violations.append(node.floatingDigits.positionAfterSkippingLeadingTrivia)
+        override func visitPost(_ node: ClassDeclSyntax) {
+            let className = node.trimmedDescription
+            if node.isXCTestCase(testParentClasses) {
+                testClasses.insert(className)
+            } else {
+                nonTestClasses.insert(className)
             }
+            processPossibleViolations(forClassName: className)
+        }
+
+        override func visitPost(_ node: FloatLiteralExprSyntax) {
+            guard node.floatingDigits.isMagicNumber else {
+                return
+            }
+            let violation = node.floatingDigits.positionAfterSkippingLeadingTrivia
+            process(violation: violation, forNode: node)
         }
 
         override func visitPost(_ node: IntegerLiteralExprSyntax) {
-            if node.isMemberOfATestClass(testParentClasses) == false, node.digits.isMagicNumber {
-                violations.append(node.digits.positionAfterSkippingLeadingTrivia)
+            guard node.digits.isMagicNumber else {
+                return
             }
+            let violation = node.digits.positionAfterSkippingLeadingTrivia
+            process(violation: violation, forNode: node)
+        }
+
+        private func process(violation: AbsolutePosition, forNode node: ExprSyntaxProtocol) {
+            guard !node.isMemberOfATestClass(testParentClasses) else {
+                return
+            }
+            if let extendedTypeName = node.extendedClassname() {
+                if testClasses.contains(extendedTypeName) {
+                    violations.append(violation)
+                } else if nonTestClasses.contains(extendedTypeName) == false {
+                    var possibleViolationsForClass = possibleViolations[extendedTypeName] ?? []
+                    possibleViolationsForClass.append(violation)
+                    possibleViolations[extendedTypeName] = possibleViolationsForClass
+                }
+            } else {
+                violations.append(violation)
+            }
+        }
+
+        private func processPossibleViolations(forClassName className: String) {
+            guard let possibleViolationsForClass = possibleViolations[className] else {
+                return
+            }
+            violations.append(contentsOf: possibleViolationsForClass)
+            violations.sort(by: { $0.position < $1.position })
+            possibleViolations[className] = nil
         }
     }
 }
@@ -125,5 +190,16 @@ private extension ExprSyntaxProtocol {
             parent = parent?.parent
         }
         return false
+    }
+
+    func extendedClassname() -> String? {
+        var parent = parent
+        while parent != nil {
+            if let extensionDecl = parent?.as(ExtensionDeclSyntax.self) {
+                return extensionDecl.extendedType.trimmedDescription
+            }
+            parent = parent?.parent
+        }
+        return nil
     }
 }

--- a/Source/SwiftLintBuiltInRules/Rules/Idiomatic/NoMagicNumbersRule.swift
+++ b/Source/SwiftLintBuiltInRules/Rules/Idiomatic/NoMagicNumbersRule.swift
@@ -133,9 +133,9 @@ private extension NoMagicNumbersRule {
             }
             let violation = node.positionAfterSkippingLeadingTrivia
             if let extendedTypeName = node.extendedTypeName() {
-                if testClasses.contains(extendedTypeName) == false {
+                if !testClasses.contains(extendedTypeName) {
                     violations.append(violation)
-                    if nonTestClasses.contains(extendedTypeName) == false {
+                    if !nonTestClasses.contains(extendedTypeName) {
                         possibleViolations[extendedTypeName, default: []].insert(violation)
                     }
                 }

--- a/Source/SwiftLintBuiltInRules/Rules/Idiomatic/NoMagicNumbersRule.swift
+++ b/Source/SwiftLintBuiltInRules/Rules/Idiomatic/NoMagicNumbersRule.swift
@@ -67,7 +67,7 @@ struct NoMagicNumbersRule: SwiftSyntaxRule, OptInRule, ConfigurationProviderRule
                    let baz: [Int] = Array(repeating: 0, count: 10)
                 }
             }
-            """).focused(),
+            """),
             Example("""
             extension FooTests {
                 class Bar {

--- a/Source/SwiftLintBuiltInRules/Rules/Idiomatic/NoMagicNumbersRule.swift
+++ b/Source/SwiftLintBuiltInRules/Rules/Idiomatic/NoMagicNumbersRule.swift
@@ -117,22 +117,21 @@ private extension NoMagicNumbersRule {
             guard node.floatingDigits.isMagicNumber else {
                 return
             }
-            let violation = node.floatingDigits.positionAfterSkippingLeadingTrivia
-            process(violation: violation, forNode: node)
+            collectViolation(forNode: node)
         }
 
         override func visitPost(_ node: IntegerLiteralExprSyntax) {
             guard node.digits.isMagicNumber else {
                 return
             }
-            let violation = node.digits.positionAfterSkippingLeadingTrivia
-            process(violation: violation, forNode: node)
+            collectViolation(forNode: node)
         }
 
-        private func process(violation: AbsolutePosition, forNode node: ExprSyntaxProtocol) {
-            guard !node.isMemberOfATestClass(testParentClasses) else {
+        private func collectViolation(forNode node: ExprSyntaxProtocol) {
+            if node.isMemberOfATestClass(testParentClasses) {
                 return
             }
+            let violation = node.positionAfterSkippingLeadingTrivia
             if let extendedTypeName = node.extendedTypeName() {
                 if testClasses.contains(extendedTypeName) == false {
                     violations.append(violation)

--- a/Source/SwiftLintCore/Protocols/SwiftSyntaxRule.swift
+++ b/Source/SwiftLintCore/Protocols/SwiftSyntaxRule.swift
@@ -138,4 +138,3 @@ public extension Set where Element == ReasonedRuleViolation {
         insert(ReasonedRuleViolation(position: position))
     }
 }
-

--- a/Source/SwiftLintCore/Protocols/SwiftSyntaxRule.swift
+++ b/Source/SwiftLintCore/Protocols/SwiftSyntaxRule.swift
@@ -86,7 +86,7 @@ public extension SwiftSyntaxRule {
 }
 
 /// A violation produced by `ViolationsSyntaxVisitor`s.
-public struct ReasonedRuleViolation: Comparable {
+public struct ReasonedRuleViolation: Comparable, Hashable {
     /// The violation's position.
     public let position: AbsolutePosition
     /// A specific reason for the violation.

--- a/Source/SwiftLintCore/Protocols/SwiftSyntaxRule.swift
+++ b/Source/SwiftLintCore/Protocols/SwiftSyntaxRule.swift
@@ -127,14 +127,3 @@ public extension Array where Element == ReasonedRuleViolation {
         append(contentsOf: positions.map { ReasonedRuleViolation(position: $0) })
     }
 }
-
-/// Extension for sets of `ReasonedRuleViolation`s that provides the automatic conversion of
-/// `AbsolutePosition`s into `ReasonedRuleViolation`s (without a specific reason).
-public extension Set where Element == ReasonedRuleViolation {
-    /// Insert a minimal violation for the specified position.
-    ///
-    /// - parameter position: The position for the violation to be inserted.
-    mutating func insert(_ position: AbsolutePosition) {
-        insert(ReasonedRuleViolation(position: position))
-    }
-}

--- a/Source/SwiftLintCore/Protocols/SwiftSyntaxRule.swift
+++ b/Source/SwiftLintCore/Protocols/SwiftSyntaxRule.swift
@@ -127,3 +127,15 @@ public extension Array where Element == ReasonedRuleViolation {
         append(contentsOf: positions.map { ReasonedRuleViolation(position: $0) })
     }
 }
+
+/// Extension for sets of `ReasonedRuleViolation`s that provides the automatic conversion of
+/// `AbsolutePosition`s into `ReasonedRuleViolation`s (without a specific reason).
+public extension Set where Element == ReasonedRuleViolation {
+    /// Insert a minimal violation for the specified position.
+    ///
+    /// - parameter position: The position for the violation to be inserted.
+    mutating func insert(_ position: AbsolutePosition) {
+        insert(ReasonedRuleViolation(position: position))
+    }
+}
+

